src/sentry/services/hybrid_cloud/rpc.py

```diff
@@ -331,6 +335,41 @@ def _create_signatures(cls) -> Mapping[str, RpcMethodSignature]:
                 model_table[base_method.__name__] = signature
         return model_table
 
+    @classmethod
+    def _get_and_validate_local_implementation(cls) -> RpcService:
+        def get_parameters(method: Callable[..., Any]) -> set[str]:
+            """Get the expected set of parameter names.
+
+            The `inspect.signature` also gives us type annotations (on parameters and
+            the return value) but it's tricky to compare those, because of
+            semantically equivalent annotations represented with unequal type tokens,
+            such as `Optional[int]` versus `int | None`.
+            """
+            sig = inspect.signature(method)
+            param_names = list(sig.parameters.keys())
+            if param_names and param_names[0] == "self":
+                del param_names[0]
+            return set(param_names)
+
+        impl = cls.get_local_implementation()
+        for method_sig in cls._get_abstract_rpc_methods():
+            method_impl = getattr(impl, method_sig.__name__)
+
+            if getattr(method_impl, "__isabstractmethod__", False):
+                raise RpcServiceSetupException(
+                    f"{type(impl).__name__} must provide a concrete implementation of `{method_sig.__name__}`"
+                )
+
+            sig_params = get_parameters(method_sig)
+            impl_params = get_parameters(method_impl)
+            if not sig_params == impl_params:
+                raise RpcServiceSetupException(
+                    f"{type(impl).__name__}.{method_sig.__name__} does not match specified parameters "
+                    f"(expected: {sig_params!r}; actual: {impl_params!r})"
+                )
```

Comment:
This is effectively just an assertion of the kind we'd usually make in a unit test.

It would be nice to do all of this validation in a unit test case, but I don't think we can rely on necessarily being able to import *all* `RpcService` subclasses into one unit test case. Doing it in live code is the most foolproof way to ensure we test every service class. It should be impossible to have this break in production without tripping it in at least one unit test.

Reply:
Doing this safety check during application startup or when a service is first used seems fine to me.